doc/design.md

```diff
@@ -184,5 +184,273 @@ spec:
+      params:
+        - name: pathToDockerFile
+          value: Dockerfile
+        - name: pathToContext
+          value: /workspace/docker-source/examples/microservices/leeroy-web #configure: may change according to your source
+      resources:
+        inputs:
+          - name: docker-source
+            resource: source-repo
+        outputs:
+          - name: builtImage
+            resource: web-image
+    - name: deploy-web
+      taskRef:
+        name: deploy-using-kubectl
+      resources:
+        inputs:
+          - name: source
+            resource: source-repo
+          - name: image
+            resource: web-image
+            from:
+              - build-skaffold-web
+      params:
+        - name: path
+          value: /workspace/source/examples/microservices/leeroy-web/kubernetes/deployment.yaml #configure: may change according to your source
+        - name: yamlPathToImage
+          value: "spec.template.spec.containers[0].image"
+```
+
+The above `Pipeline` is referencing a `Task` called `deploy-using-kubectl` defined as follows:
+
+``` yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: deploy-using-kubectl
+spec:
+  params:
+    - name: path
+      type: string
+      description: Path to the manifest to apply
+    - name: yamlPathToImage
+      type: string
+      description: |
+        The path to the image to replace in the yaml manifest (arg to yq)
+  resources:
+    inputs:
+      - name: source
+        type: git
+      - name: image
+        type: image
+  steps:
+    - name: replace-image
+      image: mikefarah/yq
+      command: ["yq"]
+      args:
+        - "w"
+        - "-i"
+        - "$(params.path)"
+        - "$(params.yamlPathToImage)"
+        - "$(resources.inputs.image.url)"
+    - name: run-kubectl
+      image: lachlanevenson/k8s-kubectl
+      command: ["kubectl"]
+      args:
+        - "apply"
+        - "-f"
+        - "$(params.path)"
+```
+
+We hope Fluid users can write the following program to express the above YAML file:
+
+``` python
+@fluid.task
+def deploy_using_kubectl(
+        source_repo: "input, git",
+        web_image: "input,image",
+        path="/workspace/source/examples/microservices/leeroy-web/kubernetes/deployment.yaml",
+        yaml_path_to_image="spec.template.spec.containers[0].image"):
+    fluid.step(image="mikefarah/yq",
+             command=["yq"],
+             args=["w",
+                   "-i",
+                   f"{path}",
+                   f"{yaml_path_to_image}",
+                   f"{image.url}"])
+    fluid.step(image="lachlanevenson/k8s-kubectl",
+             command=["kubectl"],
+             args=["apply", "-f", f"{path}"])
+
+@fluid.pipeline
+def tutorial(source_repo: "resource,git", web_image: "resource,image"):
+    build_skaffold_web = build_docker_image_from_git_source(source_repo, web_image)
+
+    deploy_web = deploy_using_kubectl(source_repo, web_image)
+    deploy_web.web_image.from(build_skaffold_web)
```

Comment:
This `deploy_web.web_image.from` syntax looks confusing. Python programmer do not do this with function parameters. 

I am afraid that this weird design might come from the fact that **a Pipeline is NOT similar to a function definition**.

Reply:
After reading more about Pipeline, I see it describes a DAG of tasks, where edges are data dependencies between tasks.

Thinking about the following example from the Tekton tutorial https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#from:

```yaml
- name: build-app
  taskRef:
    name: build-push
  resources:
    outputs:
      - name: image
        resource: my-image
- name: deploy-app
  taskRef:
    name: deploy-kubectl
  resources:
    inputs:
      - name: image
        resource: my-image
        from:
          - build-app
```

Using programming language idiom, it is simply function calls

```python
deploy_kubectl(image=build_push(my_image))
```

It seems that what we expect users to write is 

```python
@fluid.pipeline
def build_and_deploy(image):
    deploy_kubectl(image=build_push(my_image))
```

where `@fluid.pipeline` should **dry-run** the function `build_and_deploy` to analysis the function dependencies, which is `deploy_kubectl.image <- build_push`, and generate the YAML definition of the Pipeline object.

I am not sure if the above suggestion is correct, or how reasonable it is. It has been a while I haven't use Tekton.